// @progress/kendo-angular-dropdowns/dist/npm/common/filtering/filter.directive.js
/**-----------------------------------------------------------------------------------------
* Copyright © 2020 Progress Software Corporation. All rights reserved.
* Licensed under commercial license. See LICENSE.md in the project root for more information
*-------------------------------------------------------------------------------------------*/
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
var core_1 = require("@angular/core");
var filterable_base_component_1 = require("./filterable-base.component");
var util_1 = require("../util");
var DEFAULT_FILTER_SETTINGS = {
    caseSensitive: false,
    operator: 'startsWith'
};
/**
 * Implements an event handler for the `filterChange` event of a DropDowns component
 * which performs simple data filtering.
 *
 * @example
 * ```ts
 * _@Component({
 * selector: 'my-app',
 * template: `
 *  <kendo-autocomplete
 *      [data]="data"
 *      kendoDropDownFilter
 *      placeholder="e.g. Andorra">
 *  </kendo-autocomplete>
 * `
 * })
 * class AppComponent {
 *     public data: Array<string> = ["Albania", "Andorra", "Armenia", "Austria", "Azerbaijan"];
 * }
 * ```
 */
var FilterDirective = /** @class */ (function () {
    function FilterDirective(component) {
        this.component = component;
        /**
         * @hidden
         *
         * Sets whether the filtering functionality is enabled on component init.
         */
        this.filterable = true;
        this._data = [];
        this._filterSettings = DEFAULT_FILTER_SETTINGS;
    }
    Object.defineProperty(FilterDirective.prototype, "data", {
        get: function () {
            return this._data;
        },
        /**
         * The initial data that will be used as a source array for the filtering operations.
         */
        set: function (data) {
            this._data = data || [];
        },
        enumerable: true,
        configurable: true
    });
    Object.defineProperty(FilterDirective.prototype, "filterSettings", {
        get: function () {
            return this._filterSettings;
        },
        /**
         * The configuration object which sets the behavior of the `kendoDropDownFilter` directive.
         */
        set: function (settings) {
            this._filterSettings = Object.assign({}, DEFAULT_FILTER_SETTINGS, settings);
        },
        enumerable: true,
        configurable: true
    });
    FilterDirective.prototype.ngOnInit = function () {
        var _this = this;
        this.component.filterable = this.filterable;
        this.filterChangeSubscription = this.component.filterChange
            .subscribe(function (query) { return _this.component.data = _this.getFilteredData(query); });
    };
    FilterDirective.prototype.ngOnDestroy = function () {
        this.filterChangeSubscription.unsubscribe();
    };
    FilterDirective.prototype.getFilteredData = function (query) {
        var _this = this;
        var field = this.component.textField || this.component.valueField;
        return this.data.filter(function (item) { return _this.checkItem(util_1.getter(item, field), query); });
    };
    FilterDirective.prototype.checkItem = function (target, query) {
        target = this.normalizeValue(target);
        query = this.normalizeValue(query);
        return this.filterSettings.operator === 'contains' ? target.indexOf(query) !== -1 : target.indexOf(query) === 0;
    };
    FilterDirective.prototype.normalizeValue = function (value) {
        var normalizedValue = util_1.isPresent(value) ? value.toString() : '';
        return this.filterSettings.caseSensitive ? normalizedValue : normalizedValue.toLowerCase();
    };
    FilterDirective.decorators = [
        { type: core_1.Directive, args: [{
                    selector: '[kendoDropDownFilter]'
                },] },
    ];
    /** @nocollapse */
    FilterDirective.ctorParameters = function () { return [
        { type: filterable_base_component_1.FilterableDropDownComponentBase }
    ]; };
    FilterDirective.propDecorators = {
        data: [{ type: core_1.Input }],
        filterSettings: [{ type: core_1.Input, args: ['kendoDropDownFilter',] }],
        filterable: [{ type: core_1.Input }]
    };
    return FilterDirective;
}());
exports.FilterDirective = FilterDirective;
